py-polars/polars/expr/expr.py

```diff
@@ -3969,6 +3969,16 @@ def where(self, predicate: Expr) -> Self:
         """
         return self.filter(predicate)
 
+    class _map_batches_wrapper:
+        def __init__(self, function: Callable[[Series], Series | Any]):
+            self.function = function
+
+        def __call__(self, *args: Any, **kwargs: Any) -> Any:
+            result = self.function(*args, **kwargs)
+            if _check_for_numpy(result) and isinstance(result, np.ndarray):
+                result = pl.Series(result)
```

Comment:
Minor performance nits:

Can we directly return `pl.Series(result)`?

I also think we should insert the `return_dtype`  in the `pl.Series` constructor.

Reply:
Something like this? 

```python
    class _map_batches_wrapper:
        def __init__(self, function: Callable[[Series], Series | Any], return_dtype: PolarsDataType | None):
            self.function = function
            self.return_dtype = return_dtype

        def __call__(self, *args: Any, **kwargs: Any) -> Any:
            return pl.Series(self.function(*args, **kwargs), dtype=self.return_dtype)
```

Without the NumPy check, this could probably be done entirely on the Rust side, although I'm not sure how offhand.

Reply:
Yes, that but then including the numpy check.

Reply:
So just to clarify, return pl.Series(result) on the Rust side but then do the NumPy check on the Python side? 

Another option could be to simply drop the NumPy check and _always_ coerce the output to a Series (and for LazyFrame.map_batches, to a DataFrame).

Reply:
Wait a sec, if you do the pl.Series conversion on the Rust side, you can't do the NumPy check on the Python side, because it'll already be a Series by the time Python sees it. 

So I think you have to choose whether to 1) convert to Series on the Rust side, or 2) do the NumPy check. I don't think you can do both.

Reply:
No, I mean all on the python side, but we must do the numpy check. I don't want to go through the series constructor by default.

Reply:
Raising an error when the function doesn't return a Series or NumPy array breaks a bunch of tests where the function returns either a scalar or a DataFrame. Here's the code that breaks the tests:

```python
def __call__(self, *args: Any, **kwargs: Any) -> Any:
    result = self.function(*args, **kwargs)
    if not isinstance(result, pl.Series):
        if _check_for_numpy(result) and isinstance(result, np.ndarray):
            result = pl.Series(result, dtype=self.return_dtype)
        else:
            raise TypeError(
                "The function passed to map_batches() "
                "must return a Series or NumPy array"
            )
    return result
```

And here are the tests that break:

```python
_____________________________________________________________________ test_error_on_reducing_map ______________________________________________________________________
[gw6] linux -- Python 3.12.1 .../polars/.venv/bin/python
tests/unit/test_errors.py:47: in test_error_on_reducing_map
    df.select(
polars/dataframe/frame.py:8053: in select
    return self.lazy().select(*exprs, **named_exprs).collect(_eager=True)
polars/lazyframe/frame.py:1749: in collect
    return wrap_df(ldf.collect())
E   polars.exceptions.ComputeError: TypeError: The function passed to map_batches() must return a Series or NumPy array
_________________________________________________________________________ test_deadlocks_3409 _________________________________________________________________________
[gw7] linux -- Python 3.12.1 .../polars/.venv/bin/python
tests/unit/dataframe/test_df.py:3172: in test_deadlocks_3409
    assert (
polars/dataframe/frame.py:8228: in with_columns
    return self.lazy().with_columns(*exprs, **named_exprs).collect(_eager=True)
polars/lazyframe/frame.py:1749: in collect
    return wrap_df(ldf.collect())
E   polars.exceptions.ComputeError: TypeError: The function passed to map_batches() must return a Series or NumPy array
_______________________________________________________________________ test_map_batches_group ________________________________________________________________________
[gw3] linux -- Python 3.12.1 .../polars/.venv/bin/python
tests/unit/operations/map/test_map_batches.py:68: in test_map_batches_group
    assert df.group_by("id").agg(pl.col("t").map_batches(lambda s: s.sum())).sort(
polars/dataframe/group_by.py:243: in agg
    .collect(no_optimization=True)
polars/lazyframe/frame.py:1749: in collect
    return wrap_df(ldf.collect())
E   polars.exceptions.ComputeError: TypeError: The function passed to map_batches() must return a Series or NumPy array
______________________________________________________________________ test_map_return_py_object ______________________________________________________________________
[gw1] linux -- Python 3.12.1 .../polars/.venv/bin/python
tests/unit/operations/map/test_map_batches.py:15: in test_map_return_py_object
    result = df.select([pl.all().map_batches(lambda s: reduce(lambda a, b: a + b, s))])
polars/dataframe/frame.py:8053: in select
    return self.lazy().select(*exprs, **named_exprs).collect(_eager=True)
polars/lazyframe/frame.py:1749: in collect
    return wrap_df(ldf.collect())
E   polars.exceptions.ComputeError: TypeError: The function passed to map_batches() must return a Series or NumPy array
_____________________________________________________________________ test_error_on_reducing_map ______________________________________________________________________
[gw2] linux -- Python 3.12.1 .../polars/.venv/bin/python
tests/unit/operations/map/test_map_batches.py:54: in test_error_on_reducing_map
    df.select(
polars/dataframe/frame.py:8053: in select
    return self.lazy().select(*exprs, **named_exprs).collect(_eager=True)
polars/lazyframe/frame.py:1749: in collect
    return wrap_df(ldf.collect())
E   polars.exceptions.ComputeError: TypeError: The function passed to map_batches() must return a Series or NumPy array
```

Reply:
Hmm.. Ok, let's leave that for another PR to figure out then. :/

Reply:
Ok, sounds good. We should be ready to merge then.

I added the `return_dtype` argument like you requested, so now it looks like:

```python
result = self.function(*args, **kwargs)
if _check_for_numpy(result) and isinstance(result, np.ndarray):
    result = pl.Series(result, dtype=self.return_dtype)
return result
```

Note that `_check_for_numpy(result)` returns False for non-NumPy arrays and takes about the same amount of time as an isinstance check (~4e-5 seconds), so `if not isinstance(result, pl.Series) and _check_for_numpy(result) and isinstance(result, np.ndarray):` wouldn't be meaningfully faster.

Reply:
@ritchie46 good to merge?

Reply:
Yes! thanks!